lib/apiservers/engine/backends/commit.go

```diff
@@ -0,0 +1,472 @@
+		if err != nil {
+			return nil, err
+		}
+		if config.Tag != "" {
+			if imageRef, err = reference.WithTag(imageRef, config.Tag); err != nil {
+				return nil, err
+			}
+		}
+	}
+	options := imagec.Options{
+		Destination: os.TempDir(),
+		Reference:   imageRef,
+		Tag:         config.Tag,
+	}
+	portLayerServer := PortLayerServer()
+
+	if portLayerServer != "" {
+		options.Host = portLayerServer
+	}
+
+	ic := imagec.NewImageC(options, streamformatter.NewJSONStreamFormatter())
+	if imageRef != nil {
+		ic.ParseReference()
+	}
+
+	return ic, nil
+}
+
+func setLayerConfig(lm *imagec.ImageWithMeta, container *types.ContainerJSON, config *backend.ContainerCommitConfig, newConfig *containertypes.Config) error {
+	defer trace.End(trace.Begin(lm.ID))
+
+	// Host is either the host's UUID (if run on vsphere) or the hostname of
+	// the system (if run standalone)
+	host, err := sys.UUID()
+	if err != nil {
+		return InternalServerError(fmt.Sprintf("Failed to get host name: %s", err))
+	}
+
+	if host != "" {
+		log.Infof("Using UUID (%s) for imagestore name", host)
+	}
+
+	vc := cache.ContainerCache().GetContainer(container.ID)
+	meta := dockerimage.V1Image{
+		ID:              lm.ID,
+		Parent:          vc.LayerID,
+		Author:          config.Author,
+		Comment:         config.Comment,
+		Created:         time.Now().UTC(),
+		Container:       container.ID,
+		ContainerConfig: *container.Config,
+		Architecture:    runtime.GOARCH,
+		OS:              runtime.GOOS,
+		DockerVersion:   version.DockerServerVersion,
+		Config:          newConfig,
+		Size:            lm.Size,
+	}
+
+	m, err := json.Marshal(meta)
+	if err != nil {
+		return InternalServerError(fmt.Sprintf("Failed to marshal image layer config: %s", err))
+	}
+	// layer metadata
+	lm.Meta = string(m)
+	lm.Image.Parent = vc.LayerID
+	lm.Image.Store = host
+	return nil
+}
+
+func downloadDiff(rc io.ReadCloser, containerID string, options imagec.Options) (*imagec.ImageWithMeta, error) {
+	defer trace.End(trace.Begin(containerID))
+
+	// generate random string as layer ID
+	layerID := stringid.GenerateRandomID()
+
+	tmpLayerFileName, diffIDSum, gzSum, err := compressDiffToTmpFile(rc, containerID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cleanup function for the error case
+	defer func() {
+		if err != nil {
+			os.Remove(tmpLayerFileName)
+		}
+	}()
+
+	blobSum := digest.NewDigestFromBytes(digest.SHA256, gzSum)
+	log.Debugf("container %s blob sum: %s", containerID, blobSum.String())
+	diffID := digest.NewDigestFromBytes(digest.SHA256, diffIDSum)
+	log.Debugf("container %s diff id: %s", containerID, diffID.String())
+
+	layerFile, err := os.Open(string(tmpLayerFileName))
+	if err != nil {
+		return nil, err
+	}
+	defer layerFile.Close()
+
+	decompressed, err := gzip.NewReader(layerFile)
+	if err != nil {
+		return nil, err
+	}
+	defer decompressed.Close()
+
+	// get a tar reader
+	tr := tar.NewReader(decompressed)
+
+	// iterate through tar headers to get file sizes
+	var layerSize int64
```

Comment:
I don't believe you need to do all this to get the size of the uncompressed tar file.  In your compressDiffToTmpFile() function, you're doing an io.Copy.  The first return from that is bytes written for the uncompressed tar file.

I'm doing something similar in push.  We can refactor them together later.

https://github.com/sflxn/vic/blob/docker-push/lib/imagec/imagec.go#L915